### ISSUE TITLE
Question regarding json globals conv2d_same_offset_depth_1/2, etc

### ISSUE THREAD

**Issue by krworks** [Dobiasd/frugally-deep#64]
In the output from convert_model.py, I see the following:
average_pooling_2d_same_offset:false
average_pooling_2d_valid_offset:false
conv2d_same_offset_depth_1:false
conv2d_same_offset_depth_2:false
conv2d_valid_offset_depth_1:false
conv2d_valid_offset_depth_2:false
max/separable, etc

It would seem that the convert_model does some checks on the backend, sets these booleans in json. 
 These variables are then used in frugally-deep's C in places like the convolution, average_pooling, etc.  My question is, seems these depend on individual operations and/or backend used, and are 'inconsistent' as you describe in the convert_model.py, how could they be used in global form?

With CPU or GPU generation on two different models I see false for all of these.  Do you know what makes one or more of them true?

Another thing: One may take multiple inputs and perform different convolutions in stages so could you get different results inside the model?

I apologize for this but I'm just looking for clarity.

**Comment by Dobiasd** [comment]
> Do you know what makes one or more of them true?

Having certain versions of Keras/TensorFlow/Cuda/CuDNN (I don't know which of those exactly) can make some or all of them true. I introduced these things because I was confronted with system that needed them to make the output of frugally-deep equivalent to the output of Keras on the same model. Doing a convolution with of without offset does not make a real difference for the actual meaning of the result of a model, but it's important to make the automated tests succeed.

> My question is, seems these depend on individual operations and/or backend used, and are 'inconsistent' as you describe in the convert_model.py, how could they be used in global form?

What do you mean by using them in global form? As a user of frugally-deep you usually don't have to care about these booleans. :)

> One may take multiple inputs and perform different convolutions in stages so could you get different results inside the model?

Sorry, I don't understand the question. Could you please elaborate?

**Comment by krworks** [comment]
My question was could you get different booleans from different layers constructed slightly differently inside your model?  I suppose it does not matter as much for me since you say it's just for automated tests. :-)

One more silly question and I'm done: The usage of shape3() constructor seems to indicate you internally store tensors as channel first (depth always passed first) even though channels_last is the only supported import. Is that the case or am I just reading this a bit off?
convolution.hpp:
`shape3(out_depth, out_height, out_width)`

**Comment by Dobiasd** [comment]
> My question was could you get different booleans from different layers constructed slightly differently inside your model?

Ah, understood. No, at least during my tests these booleans only depended on the system, not on the layers. It was quite cumbersome to find out about these things. It started out with broken tests on one machine and working tests on another machine. From there I had to iterate down to minimal examples with so few values that I was able to calculate the convolution results by hand to revese-engineer the details of the implementations.

> The usage of shape3() constructor seems to indicate you internally store tensors as channel first (depth always passed first) even though channels_last is the only supported import. Is that the case or am I just reading this a bit off?

Absolutely correct (as stated in the README.md `frugally-deep uses channels_first (depth/channels, height, width) as its image_data_format internally` :wink:).
